### PR TITLE
8305425: Thread.isAlive0 doesn't need to call into the VM

### PR DESCRIPTION
**Pull request by shipilev**
Unclean backport to improve `Thread.isAlive` performance, and provide grounds for more backports in the area.

The majority of conflicts are due to integration of Loom in JDK 19, which reshaped some Thread code. I tried to change as little in 17u as possible to avoid the change fan-out.

Additional testing:
 - [x] Linux x86_64 fastdebug `tier1 tier2 tier3`
 - [x] Linux AArch64 fastdebug `tier1 tier2 tier3`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8305425](https://bugs.openjdk.org/browse/JDK-8305425): Thread.isAlive0 doesn't need to call into the VM (**Enhancement** - P4)

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [bf0d60c9](https://git.openjdk.org/jdk17u-dev/pull/1425/files/bf0d60c9f59669766a62a490838a77d2e45fda59)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/1425/head:pull/1425` \
`$ git checkout pull/1425`

Update a local copy of the PR: \
`$ git checkout pull/1425` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/1425/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1425`

View PR using the GUI difftool: \
`$ git pr show -t 1425`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/1425.diff">https://git.openjdk.org/jdk17u-dev/pull/1425.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/1425#issuecomment-1589881639)

### CODE REVIEW

**Review by phohensee — APPROVED** [review]
Lgtm.

File: src/hotspot/share/prims/jvm.cpp [hunk 1]
```diff
@@ -2976,11 +2976,6 @@ JVM_ENTRY(void, JVM_StopThread(JNIEnv* env, jobject jthread, jobject throwable))
 JVM_END
 
 
-JVM_ENTRY(jboolean, JVM_IsThreadAlive(JNIEnv* env, jobject jthread))
-  oop thread_oop = JNIHandles::resolve_non_null(jthread);
-  return java_lang_Thread::is_alive(thread_oop);
-JVM_END
-
```
Comment (TheRealMDoerr) [inline src/hotspot/share/prims/jvm.cpp:2979]: Seems like we should better remove one empty line more (like original change).

File: src/java.base/share/classes/java/lang/Thread.java [hunk 2]
```diff
@@ -157,7 +157,14 @@ public class Thread implements Runnable {
...
     /* Fields reserved for exclusive use by the JVM */
     private boolean stillborn = false;
-    private long eetop;
+
+    /*
+     * Reserved for exclusive use by the JVM. The historically named
+     * `eetop` holds the address of the underlying VM JavaThread, and is set to
+     * non-zero when the thread is started, and reset to zero when the thread terminates.
+     * A non-zero value indicates this thread isAlive().
+     */
```
Comment (TheRealMDoerr) [inline src/java.base/share/classes/java/lang/Thread.java:166]: Shortened comment. Ok.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. Thanks for backporting! Please consider removing an empty line.

File: src/hotspot/share/prims/jvm.cpp (hunk 1 above)
Comment (shipilev) [reply src/hotspot/share/prims/jvm.cpp]: Right! Fixed in new commit.

File: src/java.base/share/classes/java/lang/Thread.java (hunk 2 above)
Comment (shipilev) [reply src/java.base/share/classes/java/lang/Thread.java:166]: Yes. Mostly because FieldHolder is not in JDK 17, so mentioning it would be confusing, IMO.

**Review by phohensee — APPROVED** [review]
What I see in the PR:

1. Remove JVM_IsThreadAlive. Straightforward.
2. OrderAccess::release() before nulling the native thread field in ensure_join. This looks like it should always have been there, independent of this PR. Strictly speaking, there should be a storeload barrier after or as part of the call to set_thread() (which there isn't), but the store is done while the j.l.Thread object is locked, so no need.

**Comment by dholmes-ora** [comment]
@phohensee the memory ordering issues were discussed in detail in the original PR for this change in mainline.

**Comment by shipilev** [comment]
Reviews are done. Push approval is here. Testing is green.

/integrate

**Pull request closed without merging**